### DISCUSSION TITLE
How to send events between flutter and rust

### DISCUSSION THREAD

**Discussion by TibiToz** [fzyzcjy/flutter_rust_bridge#2223] [Q&A] [answered]
Hey !

What would be the simplest way of passing events between Rust and Flutter.

Rust is scanning for iot devices on my network and I would like to send async events with some data when a new device is discovered or when a device is disconnected to update accordingly my UI in flutter.

Is it possible to directly use an mpsc channel or an async::mpscchannel from tokio ?

I'm not sure how about to proceed with this.

Any ideas to get me started is very welcome :)

T.

**Comment by fzyzcjy** [comment]
Hi, what about a Stream https://cjycode.com/flutter_rust_bridge/guides/types/translatable/stream ?

**Comment by TibiToz** [comment]
I managed to send some events with stream, thank you.

But it seems that i struggle to retrieve the data in Flutter, maybe you can help me better understand the problem:

I have something like this :

```
pub struct Device {
    hostname: String,
    ip: HashSet<IpAddr>,
}


pub enum DeviceEvent {
    New(Device),
    Removed(Device),
    Connected(Device),
    Disconnected(Device),
    Error(String),
}

fn send_event ( e:  DeviceEvent<Device>) {
   stream.add(e)
}
```

On the flutter side I want to be able to match and retrieve data of specific event. But I don't understand how, even after reading the documentation.

**Comment by TibiToz** [comment, marked as answer]
I understood my mistake, I treated enum like structs.

I had to add #[frb(non_opaque)] on my enum and install freezed.

But it seems that freezed is getting a parsing error. Any ideas how to solve this issue ?

> An error `FormatterException` occurred while formatting the generated source for
>   `package:app/src/rust/api/discovery.dart`
> which was output to
>   `lib/src/rust/api/discovery.freezed.dart`.
> This may indicate an issue in the generator, the input source code, or in the
> source formatter.
> Could not format because the source could not be parsed:
> 
> line 21, column 91 of .: Expected to find ')'.
>    ╷
> 21 │ @optionalTypeArgs TResult when<TResult extends Object?>(TResult Function( Device field0)  new,{required TResult Function( Device field0)  removed,required TResult Function( Device field0)  connected,required TResult Function( Device field0)  disconnected,required TResult Function( String field0)  error,}) => throw _privateConstructorUsedError;
>    │                                                                                           ^^^
>    ╵
> line 22, column 100 of .: Expected to find ')'.
>    ╷
> 22 │ @optionalTypeArgs TResult? whenOrNull<TResult extends Object?>(TResult? Function( Device field0)?  new,{TResult? Function( Device field0)?  removed,TResult? Function( Device field0)?  connected,TResult? Function( Device field0)?  disconnected,TResult? Function( String field0)?  error,}) => throw _privateConstructorUsedError;
>    │                                                                                                    ^^^
>    ╵
> line 23, column 97 of .: Expected to find ')'.
>    ╷
> 23 │ @optionalTypeArgs TResult maybeWhen<TResult extends Object?>(TResult Function( Device field0)?  new,{TResult Function( Device field0)?  removed,TResult Function( Device field0)?  connected,TResult Function( Device field0)?  disconnected,TResult Function( String field0)?  error,required TResult orElse(),}) => throw _privateConstructorUsedError;
>    │                                                                                                 ^^^
>    ╵
> line 24, column 98 of .: Expected to find ')'.
>    ╷
> 24 │ @optionalTypeArgs TResult map<TResult extends Object?>(TResult Function( DeviceEvent_New value)  new,{required TResult Function( DeviceEvent_Removed value)  removed,required TResult Function( DeviceEvent_Connected value)  connected,required TResult Function( DeviceEvent_Disconnected value)  disconnected,required TResult Function( DeviceEvent_Error value)  error,}) => throw _privateConstructorUsedError;
>    │                                                                                                  ^^^
>    ╵
> line 25, column 107 of .: Expected to find ')'.
>    ╷
> 25 │ @optionalTypeArgs TResult? mapOrNull<TResult extends Object?>(TResult? Function( DeviceEvent_New value)?  new,{TResult? Function( DeviceEvent_Removed value)?  removed,TResult? Function( DeviceEvent_Connected value)?  connected,TResult? Function( DeviceEvent_Disconnected value)?  disconnected,TResult? Function( DeviceEvent_Error value)?  error,}) => throw _privateConstructorUsedError;
>    │                                                                                                           ^^^
>    ╵
> line 26, column 104 of .: Expected to find ')'.
>    ╷
> 26 │ @optionalTypeArgs TResult maybeMap<TResult extends Object?>(TResult Function( DeviceEvent_New value)?  new,{TResult Function( DeviceEvent_Removed value)?  removed,TResult Function( DeviceEvent_Connected value)?  connected,TResult Function( DeviceEvent_Disconnected value)?  disconnected,TResult Function( DeviceEvent_Error value)?  error,required TResult orElse(),}) => throw _privateConstructorUsedError;
>    │                                                                                                        ^^^
>    ╵
> line 112, column 91 of .: Expected to find ')'.
>     ╷
> 112 │ @optionalTypeArgs TResult when<TResult extends Object?>(TResult Function( Device field0)  new,{required TResult Function( Device field0)  removed,required TResult Function( Device field0)  connected,required TResult Function( Device field0)  disconnected,required TResult Function( String field0)  error,}) {
>     │                                                                                           ^^^
>     ╵
> line 113, column 13 of .: Expected an identifier.
>     ╷
> 113 │   return new(field0);
>     │             ^
>     ╵
> line 116, column 100 of .: Expected to find ')'.
>     ╷
> 116 │ @optionalTypeArgs TResult? whenOrNull<TResult extends Object?>(TResult? Function( Device field0)?  new,{TResult? Function( Device field0)?  removed,TResult? Function( Device field0)?  connected,TResult? Function( Device field0)?  disconnected,TResult? Function( String field0)?  error,}) {
>     │                                                                                                    ^^^
>     ╵
> line 117, column 13 of .: Expected an identifier.
>     ╷
> 117 │   return new?.call(field0);
>     │             ^^
>     ╵
> (38 more errors...)

**Reply by fzyzcjy** [reply]
Hi, could you please provide a minimal reproducible sample? Then maybe this can be debugged.

**Reply by TibiToz** [reply]
Hi, I found the problem.

New must  be a reserved keyword, I change my rust enum DeviceEvent::New to something else and it worked
